### PR TITLE
fix(dropdown): exclude calcite action from keyDownHandler 

### PR DESCRIPTION
**Pull request by Elijbet**
**Related Issue:** #3720

When button is selected with space or enter, the browser default behavior is that it triggers the click handler. If we call the click handler separately with keyboard, we end up calling it twice which then toggles it on, and immediately off, resulting in the described issue.

Here is a demo of the issue that was occurring.
demo of calling click handler twice on keyboard

Update: Trigger slot should support any slotted content, eg link, div, calcite-link, etc. Removing call for click handler on keyboard interaction removed that ability, so I added it back. Instead, I added calcite-action to a list of ignored node types in keyDownHandler. This isn’t great so there will be a follow up issue to refactor.

### CODE REVIEW

**Comment by driskull** [comment]
Nice catch. The enter key should already work with button. It was probably a div or something at one point.

File: src/components/calcite-dropdown/calcite-dropdown.e2e.ts [hunk 1]
```diff
@@ -753,6 +753,60 @@ describe("calcite-dropdown", () => {
     expect(await dropdownWrapper.isVisible()).toBe(false);
   });
 
+  describe("opens the dropdown with click, enter, or space", () => {
+    it("opens when dropdown-trigger is a button", async () => {
+      const page = await newE2EPage();
```
Comment (driskull) [inline src/components/calcite-dropdown/calcite-dropdown.e2e.ts:758]: nitpick: You can use `newE2EPage({html: "<html>"})` and then you won't have to call `setContent`.

File: src/components/calcite-dropdown/calcite-dropdown.tsx [hunk 2]
```diff
@@ -409,10 +409,6 @@ export class CalciteDropdown {
     ) {
       if (target.nodeName !== "BUTTON" && target.nodeName !== "CALCITE-BUTTON") {
```
Comment (macandcheese) [inline src/components/calcite-dropdown/calcite-dropdown.tsx:410]: Can we scope this fix to the calcite-button? As it stands this removes the ability for any other element (`calcite-link`, a `div`, etc.) to trigger the dropdown via keyboard.

File: src/components/calcite-dropdown/calcite-dropdown.e2e.ts [hunk 3]
```diff
@@ -753,6 +753,60 @@ describe("calcite-dropdown", () => {
...
+        </calcite-dropdown> 
+      `);
+      const element = await page.find("calcite-dropdown");
+      const trigger = await element.find("calcite-button[slot='dropdown-trigger']");
+      const dropdownWrapper = await page.find("calcite-dropdown >>> .calcite-dropdown-wrapper");
+
+      expect(await dropdownWrapper.isVisible()).toBe(false);
+      await trigger.click();
+      await page.waitForChanges();
+      expect(await dropdownWrapper.isVisible()).toBe(true);
```
Comment (driskull) [inline src/components/calcite-dropdown/calcite-dropdown.e2e.ts:780]: Should we also test a space/enter key here? Maybe close it and try again with space and enter key. The test name can be a bit deceiving.

Comment (driskull) [reply src/components/calcite-dropdown/calcite-dropdown.e2e.ts:780]: Maybe also test that the event is only called once each time?

**Review by driskull — CHANGES_REQUESTED** [review]
Just need to beef up the tests. 🐄

File: src/components/calcite-dropdown/calcite-dropdown.tsx (hunk 2 above)
Comment (driskull) [reply src/components/calcite-dropdown/calcite-dropdown.tsx:410]: I think this is probably out of the scope of this PR but I agree, this code looks suspicious. We shouldn't be checking node types here.

Comment (driskull) [reply src/components/calcite-dropdown/calcite-dropdown.tsx:410]: This one also: `const itemToFocus = target.nodeName !== "A" ? target : target.parentNode;`

File: src/components/calcite-dropdown/calcite-dropdown.e2e.ts (hunk 3 above)
Comment (Elijbet) [reply src/components/calcite-dropdown/calcite-dropdown.e2e.ts:781]: Added space/enter to tests + check for how many times the event has been called. Adam pointed out that trigger slot should support any slotted content, eg link, div, calcite-link, etc. Removing call for click handler on keyboard interaction removed that ability, so I added it back. Instead, I added a check for target.nodeName !== "CALCITE-ACTION” to exclude that along with the button. Matt was suggesting to get rid of checking node types, as these very feeble and can easily break if the nodeNames change. But that this can be handled in another pr. I added some demos so I can see it in action, let me know if I should remove these.

File: src/demos/calcite-dropdown.html [hunk 4]
```diff
@@ -37,6 +37,10 @@
         margin: 25px 0;
         border-top: 1px solid var(--calcite-ui-border-2);
       }
+
+      *:focus {
```
Comment (driskull) [inline src/demos/calcite-dropdown.html:41]: can you remove this styling please :)

File: src/components/calcite-dropdown/calcite-dropdown.tsx [hunk 5]
```diff
@@ -407,7 +407,11 @@ export class CalciteDropdown {
       this.triggers.includes(target) ||
       this.triggers.some((trigger) => trigger.contains(target))
     ) {
-      if (target.nodeName !== "BUTTON" && target.nodeName !== "CALCITE-BUTTON") {
```
Comment (driskull) [inline src/components/calcite-dropdown/calcite-dropdown.tsx:410]: My advice would be to not modify this as part of this PR> It seems to be out of scope. The fix isn't a feat to support action elements here.

**Review by driskull — COMMENTED** [review]
Final comments

File: src/components/calcite-dropdown/calcite-dropdown.tsx (hunk 5 above)
Comment (driskull) [reply src/components/calcite-dropdown/calcite-dropdown.tsx:410]: Ok. we can leave this code as is but can you create a follow up issue to refactor this code?

We shouldn't have to keep blacklisting elements here when a different button type of element is slotted.

We can discuss how to clean it up in the new issue.

**Review by driskull — APPROVED** [review]
👍 

Can you just remove that focus styling?

**Comment by Elijbet** [comment]
> 👍
> 
> Can you just remove that focus styling?

Yep, and modify the PR title as discussed.
